### PR TITLE
Remove config dirs from VOLUME

### PR DESCRIPTION
**Pull request by holybit**
Per docker issue 3639 Dockerfile that constitute a base image should not
set config dirs w/ VOLUME as downstream Dockerfile can not modify w/ RUN
instructions.

https://github.com/docker/docker/issues/3639

Issue #8

### CODE REVIEW

**Comment by holybit** [comment]
On freenode IRC #gocd user gchristensen said the following yesterday about `/var/go` in the container:

> but you should 100% definitely have /var/go be a volume, mounted from the host, and persisted to a hard disk which you can keep around forever
> because that is where everything important gets saved
> ... but the tricky thing is that the files in /var/go _must_ remain over time, because it is the entire data store of gocd, so you're already attached to it
> build artifacts, logs, the database, everything

I don't know if that's necessarily true. Poking around at several long running `gocd/gocd-server` containers, in one case a container that's been up six months, I don't see _anything_ be written out to `/var/go`. Thoughts?

**Comment by arvindsv** [comment]
@holybit and @gchristensen: Hoping to finish conversations around this and move forward.

Having read the whole of https://github.com/docker/docker/issues/3639, https://github.com/docker/docker/pull/7133 and https://github.com/docker/docker/issues/8647, I'm not entirely sure that using VOLUME the way it is used is a bug. No one from the docker team seems to have come down with an opinion about it. The closest I found was [this comment](https://github.com/docker/docker/issues/3639#issuecomment-144206271), which does not mention any rationale for it being that way.

However, whether it is a bug or a feature, there are problems when it is being used the way @holybit is trying to. My suggestion is that we have an -onbuild docker image, without any VOLUME instructions at all, and that can be used as a base image for any extensions. Then, the gocd-server and gocd-agent images use their corresponding -onbuild images and add the VOLUME instructions, so that they can be used by anyone, like this do today.

What do you think? That should help both use cases (a normal user's and that of someone looking to extend the base image).

**Comment by FredrikWendt** [comment]
The way I see it, the problem is really what's being put into `/etc/go` and `/var/go` - it's BOTH input and output:
- input of keystore, config/starting point, ... <- I'd like this baked into the image (immutable/phoenix server concept)
- output of updated config <- I'd like this in a VOLUME (the `/etc`parts)

If you could just separate these concerns (stable input vs "after system has booted config and workdir") - that'd clear things up for me and make it easier to use.

**Comment by arvindsv** [comment]
@FredrikWendt: I don't understand how the config and the keystore can be baked into a generic gocd/gocd-server image. These are changeable and will be changed both by a user possibly and by the server process itself. For instance, as an agent registers, the keystore and config might need to change.

I think a reasonable goal is to try and keep all changing and changeable content in a volume, so that a GoCD server image can be upgraded (15.3 to 16.2) without losing any data. Like a properly managed image is supposed to be.

The problem I see with doing that is that GoCD does not have a good concept of a data-only directory (is that what you meant by input and output)? The /var/go directory is lesser of a problem, since it is not used much, but /var/lib/go-server is a bigger problem. It has data (in db/h2db for instance), temporary files (in work/, plugins_work/, etc) and static files. Ideally, if it had the concept of a data directory, where all the data it manages is, then it would be nice. However, it doesn't, and I'm looking to see what we can do to work around it with what we have.

Is the goal I mentioned above reasonable, or am I missing anything?

**Comment by FredrikWendt** [comment]
The problem I had with keystore is to get the right X509 cert in there so that I could run the gocd server in a docker container and nothing else.
I understand there's valid use cases for pet servers with saving db and artifacts, and pet agents too. That's fine but my use case is primarily:
- build a new gocd server image, start it up as fast possible - no history to care about
- build a new gocd agent image, fire them up - no history to care about
- run 'add-and-run-seed.sh' which adds the first pipeline which adds all other pipelines using API calls

Our agents has git, mercurial, bash and docker. They're wiped away, and new agents are started, every 24 hours. (A job runs on all agents, which starts a new agent using the latest gocd agent image, and then kills itself.)

For this scenario, the keystore should only contain TLS cert used for HTTPS. All key pairs for agents are added as agents register using the correct `agentAutoRegisterKey`.

**Comment by tpbrown** [comment]
@arvindsv I probably should have read all the PRs before submitting #42.

TLDR: I _think_ I have the configuration items (including h2db) sanely exposed in a /config volume, and upgrades worked as planned.

@FredrikWendt where in a typical GoCD install does your X509 need to live?

**Comment by arvindsv** [comment]
@Tpbrown: No problem. As I said in #42, I'd make it the default and close a lot of these PRs and issues, and start over again. :)

**Comment by arvindsv** [comment]
Closing in favor of #42. That should allow this to be done in a good way?

**Comment by holybit** [comment]
@arvindsv I hope the new alpine based server will fix the issue. You did clean up the volume mounts - many thanks. I'll probably wait a few weeks before we test out the new alpine version until the issues shake out 😄

**Comment by arvindsv** [comment]
It was all @Tpbrown. I was just looking. :) Glad you think it's an improvement. I think so too. Let us know when you try it out.

**Comment by tpbrown** [comment]
@arvindsv @holybit the alpine server definitely does _not_ fix this issue :-(

I believe we will need to resort to two images.  The standard 'use it as it is' that exposes volumes that should be mounted, and a 'base image' that does not expose volumes.

**Comment by holybit** [comment]
@Tpbrown The key question is what data is being created in either `/etc/go` (i.e., configs) or `/var/go` (i.e., logs) that warrant `VOLUME` persistence? Typical use cases for both `/etc/` and `/var/` do **not** warrant a `VOLUME`. That was the point of this PR.

**Comment by tpbrown** [comment]
@holybit  Agreed on /var/go, disagree on /etc/go -- you want configurations to persist across container lifecycles.

IMO (and I'm _not_ part of the GoCD team so this is truly my opinion) you'd want config, artifacts, and logs to all be outside of the container lifecycle.    Everything else can be destroyed without losing data/metadata.

**Comment by holybit** [comment]
@arvindsv and @Tpbrown On the topic of `/etc/go` and configs in general the pattern I see most frequently in Docker Official Repositories (e.g., [Httpd](https://hub.docker.com/_/httpd/), [MySQL](https://hub.docker.com/_/mysql/), [Redis](https://hub.docker.com/_/redis/), [Postgres](https://hub.docker.com/_/postgres/), etc.) is that default config(s) are provided in the image in a fixed location that can be overridden in several ways:
- Volume mount custom config(s) to a `/etc/server.d/` location w/ `docker run -v ...` or `docker-compose`
- Add additional launch scripts w/ container run time Volume mount which an `ENTRYPOINT` script executes
- Build new image `FROM` base image w/ custom config in place of default
- etc.

The take away is that these Official Images are not concerned w/ how to persist custom config(s) - that's a devops task. Yes, the server should supply a default config and persist any data it creates. It's pretty clear if you read #21 that `VOLUME` mounting `/etc/go` or `/var/go` is a bad idea as you can't customize any configs therein.

**Comment by arvindsv** [comment]
The way I see it, there is a bit of a difference in what configuration means in Postgres vs. GoCD. In Postgres, configuration includes files like postgres.conf and pg_hba.conf, which deal with the number of workers, the ports to listen to, authentication details (md5, password, etc).

In GoCD's case, `/etc/go` has really one major file that matters, and that is `/etc/go/cruise-config.xml`. This file is more data, than config, to me. It is user-controlled and defined. GoCD will provide a default (nearly empty) `cruise-config.xml` file, if it is not provided. If Postgres had a file with a list of all DB schemas, that would be the equivalent of the file here.

IMHO, I like to think of usage scenarios rather than whether `VOLUME` is the right thing or the wrong thing. Here are some scenarios I like:
1. You're using GoCD server in a docker container, with 16.5 and you want to upgrade to 16.6 and keep all your "data". That makes sense to me. For this to happen, `/etc/go/cruise-config.xml` cannot be destroyed during this upgrade. @Tpbrown's changes do allow this, without using `VOLUME` (but I would have been quite happy if it did, as long as it suited this use case).
2. Similar to use case 1, you want to move your GoCD server from one Docker host to another. This should be easy.
3. I want to use the GoCD server image, but I want to install extra software by default (this is where we start getting into the Dockerfile a bit).

I can't think of any other case at this point. I felt that @Tpbrown's changes allowed for all three. I am not agreeing with either one of you in this discussion. :) I just wanted to say that making it easy for people to use the image in different ways is IMO better than trying to understand how docker expects `VOLUME` to be used.

**Comment by arvindsv** [comment]
Maybe I slid too much to the pragmatism side. I don't know.

**Comment by arvindsv** [comment]
Also, @Tpbrown's current Dockerfile doesn't expose either `/var/go` or `/etc/go` directly. It symlinks to them.
